Add engineGetModelTextures function

Lpsd opened the pull request:
Resolves #1040 (although implementation is different from suggested)

This function allows you to get the textures of any model.

`engineGetModelTextures(model)`

**Arguments:**
model: either the model ID or model name

**Returns:**
Returns a table of texture elements _[textureName, texture]_, false otherwise.

**Example:**
_Get the textures for model ID 3722 and draw them with dxDrawImage_
```lua
function init()
	textures = engineGetModelTextures(3722)
	addEventHandler("onClientRender", root, render)
end
addEventHandler("onClientResourceStart", resourceRoot, init)

function render()
	local offset = 0
	for name,texture in pairs(textures) do
		local size = dxGetPixelsSize(dxGetTexturePixels(texture))
		dxDrawImage(0+offset, 0, size, size, texture)
		offset = offset + size
	end
end
```
![Example code](https://user-images.githubusercontent.com/40902730/63220490-a3d30100-c180-11e9-9819-a32c8fe1109d.png)

Lpsd committed:
Implement engineGetModelTextures

Lpsd committed:
Include texture name in returned table

Lpsd committed:
Minor cleanup

sbx320 commented on Client/game_sa/CRenderWareSA.cpp:
```diff
@@ -755,6 +756,56 @@ void CRenderWareSA::GetModelTextureNames(std::vector<SString>& outNameList, usho
+    outTextureList.clear();
+
+    // Special case for CJ
+    //if (usModelId == 0)
```
Is this intentionally commented out?

sbx320 commented on Client/game_sa/CRenderWareSA.cpp:
```diff
@@ -755,6 +756,56 @@ void CRenderWareSA::GetModelTextureNames(std::vector<SString>& outNameList, usho
+    std::vector<RwTexture*> rwTextureList;
+    GetTxdTextures(rwTextureList, pTXD);
+
+    for (std::vector<RwTexture*>::iterator iter = rwTextureList.begin(); iter != rwTextureList.end(); iter++)
```
Just use a range-based for here for readability

sbx320 commented on Client/game_sa/CRenderWareSA.cpp:
```diff
@@ -755,6 +756,56 @@ void CRenderWareSA::GetModelTextureNames(std::vector<SString>& outNameList, usho
+        RwD3D9Raster* pD3DRaster = (RwD3D9Raster*)(&(*iter)->raster->renderResource);
+        CPixels       texture;
+        g_pCore->GetGraphics()->GetPixelsManager()->GetTexturePixels(pD3DRaster->texture, texture);
+        outTextureList.emplace_back((*iter)->name, texture);
```
Using `outTextureList.emplace_back((*iter)->name, std::move(texture));` here would allow some  further compiler optimizations by using moves.

sbx320 commented on Client/mods/deathmatch/logic/luadefs/CLuaEngineDefs.cpp:
```diff
@@ -928,6 +930,40 @@ int CLuaEngineDefs::EngineGetVisibleTextureNames(lua_State* luaVM)
+        g_pGame->GetRenderWare()->GetModelTextures(textureList, usModelID);
+
+        lua_newtable(luaVM);
+        for (auto pair : textureList)
```
should be `const auto&`

sbx320 commented on Client/mods/deathmatch/logic/luadefs/CLuaEngineDefs.cpp:
```diff
@@ -928,6 +930,40 @@ int CLuaEngineDefs::EngineGetVisibleTextureNames(lua_State* luaVM)
+        for (auto pair : textureList)
+        {
+            CClientTexture* pTexture = g_pClientGame->GetManager()->GetRenderElementManager()->CreateTexture("", &std::get<1>(pair), RDEFAULT, RDEFAULT,RDEFAULT, RFORMAT_UNKNOWN, TADDRESS_WRAP);
+            lua_pushstring(luaVM, (const char*)&std::get<0>(pair));
```
This does not look correct. Just do `std::get<0>(pair).c_str()` here instead of weird casts.

sbx320 commented on Client/mods/deathmatch/logic/luadefs/CLuaEngineDefs.cpp:
```diff
@@ -928,6 +930,40 @@ int CLuaEngineDefs::EngineGetVisibleTextureNames(lua_State* luaVM)
+        return luaL_error(luaVM, argStream.GetFullErrorMessage());
+
+    ushort usModelID = CModelNames::ResolveModelID(strModelName);
+    if (usModelID != INVALID_MODEL_ID)
```
I'd prefer `if (usModelID == INVALID_MODEL_ID) return 1;` here to reduce indentation depth and to keep consistent with the `HasErrors` call above.

sbx320 reviewed (dismissed):
Looks good, added some minor style comments. I'm not 100% sure on the lifetimes of the textures though. Might be worth trying to call engineGetModelTextures a few thousand times and seeing what happens to memory usage.

CrosRoad95 commented:
does it work for custom models? you should also add filter parametr - which texture you want in case models has many, high quality textures and you want just one.

you should also do test by loop all models 1 to 20000 and check if it crash

Lpsd committed:
Implement suggested changes

Lpsd commented:
@sbx320 I have implemented the suggested changes, aside from the "range-based for loop" in CRenderWareSA - this seems to crash the client and I can't figure out why.

I have tested all model ID's to 20000 and there is no crash, also as long as you `destroyElement` on the texture then you are fine, no memory leak.

Next I'll implement "filter" parameter suggested by @CrosRoad95 , so you can get specific texture(s) from the result instead of returning all. After that I think this PR is ready.

Lpsd committed:
Add textureNames parameter

allows you to select which textures you want from the model
2 usage examples
engineGetModelTextures(100, "texture_name")
engineGetModelTextures(100, {"texture_name", "another_texture_name"})

Lpsd commented:
Added `textureNames` parameter. Allows you to choose which textures you want from the model by name. Supports a single string or a table of strings. Examples:

```lua
engineGetModelTextures(3722, "Was_sd")
engineGetModelTextures(3722, {"Was_sd", "Was_side"})
```

Lpsd committed:
Small formatting fix

forkerer commented:
I would suggest using pattern matching similar to the one used in `engineApplyShaderToWorldTexture`, that supports wildcards and is already used in codebase

JobiwanD commented:
For the "range-based loop", I've changed these lines and it works well.

```C++
for (RwTexture* pTexture : rwTextureList)
{
    SString strTextureName = pTexture->name;

    /* ... */

    if (bValidTexture)
    {
        RwD3D9Raster* pD3DRaster = (RwD3D9Raster*)(&pTexture->raster->renderResource);
        /* ... */
    }

    /* ... */
}
```

I have restarted many times (~5000) a resource containing an improved version of your example. The memory usage remains stable **but**, as you have explained, the resource has to call `destroyElement` for each texture and for the table itself.

```lua
function init()
	textures = engineGetModelTextures(3722)
	addEventHandler("onClientRender", root, render)
end
addEventHandler("onClientResourceStart", resourceRoot, init)

function terminate()
	for name,texture in pairs(textures) do
		destroyElement(texture)
	end
	destroyElement(textures)
end
addEventHandler("onClientResourceStop", resourceRoot, terminate)

function render()
	local offset = 0
	for name,texture in pairs(textures) do
		local pixels = dxGetTexturePixels(texture)
		local size = dxGetPixelsSize(pixels)
		dxDrawImage(0+offset, 0, size, size, texture)
		offset = offset + size
	end
end

```

JobiwanD commented:
@CrosRoad95 if you try it with a custom model, you will get the textures of the original model and the imported ones.

![image](https://user-images.githubusercontent.com/29665323/65834630-55fffb80-e2dd-11e9-8385-9fae2b8bc2dc.png)

How can I help for the pattern part?

Lpsd committed:
Add suggested changes

range-based loop and wildcard matching

Lpsd commented:
I have implemented the suggestion for wildcard matching and cleaned up the ugly loop in `CRenderWareSA::GetModelTextures`

Requesting review

qaisjp commented on Client/mods/deathmatch/logic/luadefs/CLuaEngineDefs.cpp:
```diff
@@ -1068,6 +1070,49 @@ int CLuaEngineDefs::EngineGetVisibleTextureNames(lua_State* luaVM)
+    ushort usModelID = CModelNames::ResolveModelID(strModelName);
+    if (usModelID == INVALID_MODEL_ID)
+    {
+        lua_pushboolean(luaVM, false);
```
print warning here?

qaisjp commented on Client/mods/deathmatch/logic/luadefs/CLuaEngineDefs.cpp:
```diff
@@ -1068,6 +1070,49 @@ int CLuaEngineDefs::EngineGetVisibleTextureNames(lua_State* luaVM)
+    }
+
+    std::vector<std::tuple<std::string, CPixels>> textureList;
+    g_pGame->GetRenderWare()->GetModelTextures(textureList, usModelID, vTextureNames);
```
so your `GetModelTextures` will simply leave `vTextureNames` empty if the texture names do not exist.

have you considered returning a bool, and returning false in failure scenarios? just a thought.

qaisjp reviewed (approved):
Requested a change or two but looks good

Lpsd committed:
Merge branch 'master' into gametexture

Lpsd committed:
Return errors on invalid model ID

Lpsd commented:
The requested changes have been made.
